File: contrib/python-bindings/source/tria_accessor_wrapper.cc
// ------------------------------------------------------------------------
//
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright (C) 2019 - 2023 by the deal.II authors
//
// This file is part of the deal.II library.
//
// Part of the source code is dual licensed under Apache-2.0 WITH
// LLVM-exception OR LGPL-2.1-or-later. Detailed license information
// governing the source code and code contributions can be found in
// LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
//
// ------------------------------------------------------------------------

#include <boost/python.hpp>

#include <point_wrapper.h>
#include <tria_accessor_wrapper.h>
#include <triangulation_wrapper.h>

DEAL_II_NAMESPACE_OPEN

namespace python
{
  namespace internal
  {
    template <int structdim, int dim, int spacedim>
    PointWrapper
    get_barycenter(const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      Point<spacedim>     barycenter = accessor->barycenter();
      boost::python::list barycenter_list;
      for (int i = 0; i < spacedim; ++i)
        barycenter_list.append(barycenter[i]);

      return PointWrapper(barycenter_list);
    }



    template <int structdim, int dim, int spacedim>
    PointWrapper
    get_center(const bool  respect_manifold,
               const bool  interpolate_from_surrounding,
               const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      Point<spacedim> center =
        accessor->center(respect_manifold, interpolate_from_surrounding);
      boost::python::list center_list;
      for (int i = 0; i < spacedim; ++i)
        center_list.append(center[i]);

      return PointWrapper(center_list);
    }



    template <int structdim, int dim, int spacedim>
    void
    set_boundary_id(const int boundary_id, void *tria_accessor)
    {
      TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<TriaAccessor<structdim, dim, spacedim> *>(tria_accessor);
      accessor->set_boundary_id(boundary_id);
    }



    template <int structdim, int dim, int spacedim>
    int
    get_boundary_id(const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);

      return accessor->boundary_id();
    }



    template <int structdim, int dim, int spacedim>
    void
    set_all_boundary_ids(const int boundary_id, void *tria_accessor)
    {
      TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<TriaAccessor<structdim, dim, spacedim> *>(tria_accessor);
      accessor->set_all_boundary_ids(boundary_id);
    }


    template <int structdim, int dim, int spacedim>
    void
    set_vertex(const int i, PointWrapper &point_wrapper, void *tria_accessor)
    {
      TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<TriaAccessor<structdim, dim, spacedim> *>(tria_accessor);
      Point<spacedim> *point =
        static_cast<Point<spacedim> *>(point_wrapper.get_point());

      accessor->vertex(i) = *point;
    }



    template <int structdim, int dim, int spacedim>
    PointWrapper
    get_vertex(const int i, const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      Point<spacedim> vertex = accessor->vertex(i);

      boost::python::list coordinates;
      for (int i = 0; i < spacedim; ++i)
        coordinates.append(vertex[i]);

      return PointWrapper(coordinates);
    }



    template <int structdim, int dim, int spacedim>
    void
    set_manifold_id(const int manifold_id, void *tria_accessor)
    {
      TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<TriaAccessor<structdim, dim, spacedim> *>(tria_accessor);
      accessor->set_manifold_id(manifold_id);
    }



    template <int structdim, int dim, int spacedim>
    int
    get_manifold_id(const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      return accessor->manifold_id();
    }



    template <int structdim, int dim, int spacedim>
    bool
    at_boundary(const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      return accessor->at_boundary();
    }


    template <int structdim, int dim, int spacedim>
    double
    measure(const void *tria_accessor)
    {
      const TriaAccessor<structdim, dim, spacedim> *accessor =
        static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
          tria_accessor);
      return accessor->measure();
    }

    template <int structdim, int dim, int spacedim>
    const TriaAccessor<structdim, dim, spacedim> *
    tria_accessor_cast(const void *tria_accessor)
    {
      return static_cast<const TriaAccessor<structdim, dim, spacedim> *>(
        tria_accessor);
    }
  } // namespace internal



  TriaAccessorWrapper::TriaAccessorWrapper(const TriaAccessorWrapper &other)
    : structdim(other.structdim)
    , dim(other.dim)
    , spacedim(other.spacedim)
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      {
        TriaAccessor<1, 2, 2> *other_accessor =
          static_cast<TriaAccessor<1, 2, 2> *>(other.tria_accessor);
        tria_accessor = new TriaAccessor<1, 2, 2>(*other_accessor);
      }
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      {
        TriaAccessor<1, 2, 3> *other_accessor =
          static_cast<TriaAccessor<1, 2, 3> *>(other.tria_accessor);
        tria_accessor = new TriaAccessor<1, 2, 3>(*other_accessor);
      }
    else if ((dim == 3) && (spacedim == 3) && (structdim == 2))
      {
        TriaAccessor<2, 3, 3> *other_accessor =
          static_cast<TriaAccessor<2, 3, 3> *>(other.tria_accessor);
        tria_accessor = new TriaAccessor<2, 3, 3>(*other_accessor);
      }
    else
      AssertThrow(false,
                  ExcMessage("Wrong structdim-dim-spacedim combination."));
  }

  TriaAccessorWrapper::TriaAccessorWrapper(void     *tria_accessor,
                                           const int structdim,
                                           const int dim,
                                           const int spacedim)
    : structdim(structdim)
    , dim(dim)
    , spacedim(spacedim)
    , tria_accessor(tria_accessor)
  {}


  TriaAccessorWrapper::~TriaAccessorWrapper()
  {
    if (dim != -1)
      {
        if ((dim == 2) && (spacedim == 2) && (structdim == 1))
          {
            // We cannot call delete on a void pointer so cast the void pointer
            // back first.
            TriaAccessor<1, 2, 2> *tmp =
              static_cast<TriaAccessor<1, 2, 2> *>(tria_accessor);
            delete tmp;
          }
        else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
          {
            TriaAccessor<1, 2, 3> *tmp =
              static_cast<TriaAccessor<1, 2, 3> *>(tria_accessor);
            delete tmp;
          }
        else
          {
            TriaAccessor<2, 3, 3> *tmp =
              static_cast<TriaAccessor<2, 3, 3> *>(tria_accessor);
            delete tmp;
          }

        dim           = -1;
        spacedim      = -1;
        structdim     = -1;
        tria_accessor = nullptr;
      }
  }



  PointWrapper
  TriaAccessorWrapper::get_barycenter() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::get_barycenter<1, 2, 2>(tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::get_barycenter<1, 2, 3>(tria_accessor);
    else
      return internal::get_barycenter<2, 3, 3>(tria_accessor);
  }



  PointWrapper
  TriaAccessorWrapper::get_center(const bool respect_manifold,
                                  const bool interpolate_from_surrounding) const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::get_center<1, 2, 2>(respect_manifold,
                                           interpolate_from_surrounding,
                                           tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::get_center<1, 2, 3>(respect_manifold,
                                           interpolate_from_surrounding,
                                           tria_accessor);
    else
      return internal::get_center<2, 3, 3>(respect_manifold,
                                           interpolate_from_surrounding,
                                           tria_accessor);
  }



  void
  TriaAccessorWrapper::set_boundary_id(const int boundary_id)
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      internal::set_boundary_id<1, 2, 2>(boundary_id, tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      internal::set_boundary_id<1, 2, 3>(boundary_id, tria_accessor);
    else
      internal::set_boundary_id<2, 3, 3>(boundary_id, tria_accessor);
  }



  void
  TriaAccessorWrapper::set_all_boundary_ids(const int boundary_id)
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      internal::set_all_boundary_ids<1, 2, 2>(boundary_id, tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      internal::set_all_boundary_ids<1, 2, 3>(boundary_id, tria_accessor);
    else
      internal::set_all_boundary_ids<2, 3, 3>(boundary_id, tria_accessor);
  }



  int
  TriaAccessorWrapper::get_boundary_id() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::get_boundary_id<1, 2, 2>(tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::get_boundary_id<1, 2, 3>(tria_accessor);
    else
      return internal::get_boundary_id<2, 3, 3>(tria_accessor);
  }



  void
  TriaAccessorWrapper::set_vertex(const int i, PointWrapper &point_wrapper)
  {
    AssertThrow(i < static_cast<int>(Utilities::pow(2, dim)),
                ExcVertexDoesNotExist(i, Utilities::pow(2, dim)));
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      internal::set_vertex<1, 2, 2>(i, point_wrapper, tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      internal::set_vertex<1, 2, 3>(i, point_wrapper, tria_accessor);
    else
      internal::set_vertex<2, 3, 3>(i, point_wrapper, tria_accessor);
  }



  PointWrapper
  TriaAccessorWrapper::get_vertex(const int i) const
  {
    AssertThrow(i < static_cast<int>(Utilities::pow(2, dim)),
                ExcVertexDoesNotExist(i, Utilities::pow(2, dim)));
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::get_vertex<1, 2, 2>(i, tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::get_vertex<1, 2, 3>(i, tria_accessor);
    else
      return internal::get_vertex<2, 3, 3>(i, tria_accessor);
  }



  void
  TriaAccessorWrapper::set_manifold_id(const int manifold_id)
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      internal::set_manifold_id<1, 2, 2>(manifold_id, tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      internal::set_manifold_id<1, 2, 3>(manifold_id, tria_accessor);
    else
      internal::set_manifold_id<2, 3, 3>(manifold_id, tria_accessor);
  }



  int
  TriaAccessorWrapper::get_manifold_id() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::get_manifold_id<1, 2, 2>(tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::get_manifold_id<1, 2, 3>(tria_accessor);
    else
      return internal::get_manifold_id<2, 3, 3>(tria_accessor);
  }



  bool
  TriaAccessorWrapper::at_boundary() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::at_boundary<1, 2, 2>(tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::at_boundary<1, 2, 3>(tria_accessor);
    else
      return internal::at_boundary<2, 3, 3>(tria_accessor);
  }



  double
  TriaAccessorWrapper::measure() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::measure<1, 2, 2>(tria_accessor);
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::measure<1, 2, 3>(tria_accessor);
    else
      return internal::measure<2, 3, 3>(tria_accessor);
  }



  unsigned int
  TriaAccessorWrapper::n_vertices() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::tria_accessor_cast<1, 2, 2>(tria_accessor)->n_vertices();
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::tria_accessor_cast<1, 2, 3>(tria_accessor)->n_vertices();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 1))
      return internal::tria_accessor_cast<1, 3, 3>(tria_accessor)->n_vertices();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 2))
      return internal::tria_accessor_cast<2, 3, 3>(tria_accessor)->n_vertices();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 3))
      return internal::tria_accessor_cast<3, 3, 3>(tria_accessor)->n_vertices();
    else
      {
        AssertThrow(false,
                    ExcMessage("Wrong structdim-dim-spacedim combination."));
        return -1;
      }
  }



  unsigned int
  TriaAccessorWrapper::n_lines() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::tria_accessor_cast<1, 2, 2>(tria_accessor)->n_lines();
    else if ((dim == 2) && (spacedim == 3) && (structdim == 1))
      return internal::tria_accessor_cast<1, 2, 3>(tria_accessor)->n_lines();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 1))
      return internal::tria_accessor_cast<1, 3, 3>(tria_accessor)->n_lines();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 2))
      return internal::tria_accessor_cast<2, 3, 3>(tria_accessor)->n_lines();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 3))
      return internal::tria_accessor_cast<3, 3, 3>(tria_accessor)->n_lines();
    else
      {
        AssertThrow(false,
                    ExcMessage("Wrong structdim-dim-spacedim combination."));
        return -1;
      }
  }



  unsigned int
  TriaAccessorWrapper::n_faces() const
  {
    if ((dim == 2) && (spacedim == 2) && (structdim == 1))
      return internal::tria_accessor_cast<1, 2, 2>(tria_accessor)->n_faces();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 2))
      return internal::tria_accessor_cast<2, 3, 3>(tria_accessor)->n_faces();
    else if ((dim == 3) && (spacedim == 3) && (structdim == 3))
      return internal::tria_accessor_cast<3, 3, 3>(tria_accessor)->n_faces();
    else
      {
        AssertThrow(false,
                    ExcMessage("Wrong structdim-dim-spacedim combination."));
        return -1;
      }
  }

} // namespace python

DEAL_II_NAMESPACE_CLOSE

File: contrib/python-bindings/tests/tria_accessor_wrapper.py
## -----------------------------------------------------------------------------
##
## SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
## Copyright (C) 2024 - 2025 by the deal.II authors
##
## This file is part of the deal.II library.
##
## Detailed license information governing the source code and contributions
## can be found in LICENSE.md and CONTRIBUTING.md at the top level directory.
##
## -----------------------------------------------------------------------------

import math
import unittest

try:
    from PyDealII.Debug import *
except ImportError:
    from PyDealII.Release import *


class TestTriaAccessorWrapper2D(unittest.TestCase):
    def setUp(self):
        self.triangulation = Triangulation("2D")
        self.triangulation.generate_hyper_cube()
        self.triangulation.refine_global(1)

    def test_at_boundary(self):
        # On a once-refined unit square, boundary faces are on the outer edges.
        # Count boundary faces and interior faces separately.
        n_boundary = 0
        n_interior = 0
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    n_boundary += 1
                else:
                    n_interior += 1
        # 4 cells, each with 4 faces = 16 face-slots.
        # 8 boundary half-edges + 4 interior edges counted twice (once per cell) = 8+8=16.
        self.assertEqual(n_boundary, 8)
        self.assertEqual(n_interior, 8)

    def test_boundary_id(self):
        # By default all boundary faces have boundary_id 0. Verify, then change.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    self.assertEqual(face.boundary_id, 0)

        # Set each boundary face to a distinct id.
        bid = 1
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    face.boundary_id = bid
                    bid += 1

        # Verify that no boundary face still has id 0.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    self.assertNotEqual(face.boundary_id, 0)

    def test_manifold_id(self):
        mid = 5
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                face.manifold_id = mid
                self.assertEqual(face.manifold_id, mid)
                mid += 1

    def test_set_all_boundary_ids(self):
        # set_all_boundary_ids propagates the id to all sub-objects of the face.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    face.set_all_boundary_ids(42)
                    self.assertEqual(face.boundary_id, 42)

    def test_measure(self):
        # After one global refinement the unit square contains 4 cells of side
        # length 0.5, so each face (edge) has measure 0.5.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                self.assertAlmostEqual(face.measure(), 0.5)

        # Total boundary length must equal the perimeter of the unit square.
        total_boundary_length = 0.0
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    total_boundary_length += face.measure()
        self.assertAlmostEqual(total_boundary_length, 4.0)

    def test_barycenter(self):
        # Verify that face barycenters lie on the unit-square boundary or interior.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                bc = face.barycenter()
                self.assertTrue(0.0 <= bc.x <= 1.0)
                self.assertTrue(0.0 <= bc.y <= 1.0)

    def test_center_matches_barycenter_for_flat_faces(self):
        # For flat (straight) faces center() and barycenter() must agree.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                bc = face.barycenter()
                c = face.center()
                self.assertAlmostEqual(bc.x, c.x)
                self.assertAlmostEqual(bc.y, c.y)

    def test_get_vertex(self):
        # Each 2-D face (edge) has exactly 2 vertices.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                v0 = face.get_vertex(0)
                v1 = face.get_vertex(1)
                # Both vertices must be inside (or on the boundary of) [0, 1]^2.
                for v in (v0, v1):
                    self.assertTrue(0.0 <= v.x <= 1.0)
                    self.assertTrue(0.0 <= v.y <= 1.0)
                # The two endpoints must be distinct.
                self.assertFalse(v0.x == v1.x and v0.y == v1.y)

    def test_set_vertex(self):
        # Move one vertex of the very first face and verify it changed.
        for cell in self.triangulation.active_cells():
            face = cell.faces()[0]
            original = face.get_vertex(0)
            new_point = Point([original.x + 0.05, original.y + 0.05])
            face.set_vertex(0, new_point)
            moved = face.get_vertex(0)
            self.assertAlmostEqual(moved.x, new_point.x)
            self.assertAlmostEqual(moved.y, new_point.y)
            break


class TestTriaAccessorWrapper3D(unittest.TestCase):
    def setUp(self):
        self.triangulation = Triangulation("3D")
        self.triangulation.generate_hyper_cube()

    def test_at_boundary(self):
        # An unrefined unit cube has 1 cell with 6 boundary faces.
        n_boundary = 0
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    n_boundary += 1
        self.assertEqual(n_boundary, 6)

    def test_boundary_id(self):
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                if face.at_boundary():
                    self.assertEqual(face.boundary_id, 0)
                    face.boundary_id = 7
                    self.assertEqual(face.boundary_id, 7)

    def test_manifold_id(self):
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                face.manifold_id = 3
                self.assertEqual(face.manifold_id, 3)

    def test_measure(self):
        # Each face of the unrefined unit cube is a unit square with area 1.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                self.assertAlmostEqual(face.measure(), 1.0)

    def test_measure_after_refinement(self):
        # After one global refinement each face has area 0.25.
        self.triangulation.refine_global(1)
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                self.assertAlmostEqual(face.measure(), 0.25)

    def test_barycenter(self):
        for cell in self.triangulation.active_cells():
            bc = cell.barycenter()
            self.assertTrue(0.0 <= bc.x <= 1.0)
            self.assertTrue(0.0 <= bc.y <= 1.0)
            self.assertTrue(0.0 <= bc.z <= 1.0)

    def test_get_vertex(self):
        # Each 3-D face (quad) has exactly 4 vertices.
        for cell in self.triangulation.active_cells():
            for face in cell.faces():
                vertices = [face.get_vertex(i) for i in range(4)]
                for v in vertices:
                    self.assertTrue(0.0 <= v.x <= 1.0)
                    self.assertTrue(0.0 <= v.y <= 1.0)
                    self.assertTrue(0.0 <= v.z <= 1.0)


if __name__ == "__main__":
    unittest.main()
